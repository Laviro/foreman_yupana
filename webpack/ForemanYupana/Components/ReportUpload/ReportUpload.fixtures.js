export const status = 'some-status';
export const processID = 0;
