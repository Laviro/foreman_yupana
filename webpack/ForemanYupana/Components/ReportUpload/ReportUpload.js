import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'patternfly-react';
import TabContainer from '../TabContainer';
import TabHeader from '../TabHeader';
import TabFooter from '../TabFooter';
import TabBody from '../TabBody';
import Tree from '../Tree';
import FileDownload from '../FileDownload';
import './reportUpload.scss';

const ReportUpload = ({
  exitCode,
  loading,
  logs,
  completed,
  files,
  downloadReports,
  restartProcess,
  error,
}) => (
  <TabContainer className="report-upload">
    <TabHeader exitCode={exitCode} onRestart={restartProcess} />
    <TabBody
      loading={loading}
      logs={logs}
      completed={completed}
      error={error}
    />
    <TabFooter>
      <Tree files={files} />
      <FileDownload onClick={downloadReports} />
    </TabFooter>
  </TabContainer>
);

ReportUpload.propTypes = {
  exitCode: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  loading: PropTypes.bool,
  logs: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]),
  completed: PropTypes.number,
  restartProcess: PropTypes.func,
  files: PropTypes.arrayOf(PropTypes.string),
  downloadReports: PropTypes.func,
  error: PropTypes.string,
};

ReportUpload.defaultProps = {
  exitCode: 0,
  loading: false,
  logs: ['No running process'],
  completed: 0,
  restartProcess: noop,
  files: [],
  downloadReports: noop,
  error: null,
};

export default ReportUpload;
