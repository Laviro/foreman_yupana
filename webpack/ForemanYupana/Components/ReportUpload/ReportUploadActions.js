import {
  REPORT_UPLOAD_PROCESS_START,
  REPORT_UPLOAD_PROCESS_STOP,
  REPORT_UPLOAD_PROCESS_FINISH,
  REPORT_UPLOAD_PROCESS_RESTART,
  REPORT_UPLOAD_FILES_DOWNLOAD,
} from './ReportUploadConstants';
import { seperator, addLogs } from '../Dashboard/DashboardHelper';

const addUploadingLogs = logs => addLogs('uploading', logs);

export const startProcess = () => dispatch => {
  // TODO: Add API call here
  const processID = setInterval(() => {
    const mockLogs = [
      'Uploading...',
      'Reports: 10',
      seperator,
      'writing report 1/10',
    ];

    const { uploading } = window.__yupana__;
    const getLastLog = () => uploading.logs[uploading.logs.length - 1];
    if (getLastLog() === seperator) {
      addUploadingLogs(mockLogs);
    }
    const splittedWords = getLastLog().split(' ');
    const lastWord = splittedWords.pop();
    let amount;
    let total;
    if (lastWord.includes('/')) {
      const splittedAmounts = lastWord.split('/');
      amount = Number(splittedAmounts[0]);
      total = Number(splittedAmounts[1]);
    }
    if (total > 0 && amount === total) {
      clearInterval(processID);
      addUploadingLogs(['Done!', seperator]);
      dispatch({
        type: REPORT_UPLOAD_PROCESS_FINISH,
        payload: { status: 'success' },
      });
      return;
    }
    const nextAmount = amount + 1;
    const newLog = `${splittedWords[0]} ${
      splittedWords[1]
    } ${nextAmount}/${total}`;
    addUploadingLogs([newLog]);
    uploading.completed = parseFloat(((nextAmount * 100) / total).toFixed(2));
    uploading.files = ['213783213', '213213213', '101763276', '12387892712'];
  }, 3000);

  dispatch({
    type: REPORT_UPLOAD_PROCESS_START,
    payload: {
      processID,
      status: 'running',
    },
  });
};

export const stopProcess = processID => dispatch => {
  // TODO: Add API call here
  clearInterval(processID);
  addUploadingLogs(['No running process', seperator]);
  dispatch({
    type: REPORT_UPLOAD_PROCESS_STOP,
    payload: { status: 'stopped' },
  });
};

export const restartProcess = () => {
  addUploadingLogs(['Restarting', seperator]);
  return {
    type: REPORT_UPLOAD_PROCESS_RESTART,
  };
};

export const downloadReports = () => dispatch => {
  // TODO: Add API call here
  setTimeout(() => {
    const response = {
      file:
        'https://download.fedoraproject.org/pub/fedora/linux/releases/30/Workstation/x86_64/iso/Fedora-Workstation-netinst-x86_64-30-1.2.iso',
    };
    // server sent the url to the file!
    // now, let's download:
    window.location.href = response.file;
  }, 100);
  dispatch({
    type: REPORT_UPLOAD_FILES_DOWNLOAD,
  });
};
