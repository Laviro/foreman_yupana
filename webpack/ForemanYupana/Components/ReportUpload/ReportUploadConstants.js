export const REPORT_UPLOAD_PROCESS_START = 'YUPANA/REPORT_UPLOAD_PROCESS_START';
export const REPORT_UPLOAD_PROCESS_STOP = 'YUPANA/REPORT_UPLOAD_PROCESS_STOP';
export const REPORT_UPLOAD_PROCESS_FINISH =
  'YUPANA/REPORT_UPLOAD_PROCESS_FINISH';
export const REPORT_UPLOAD_PROCESS_RESTART =
  'YUPANA/REPORT_UPLOAD_PROCESS_RESTART';
