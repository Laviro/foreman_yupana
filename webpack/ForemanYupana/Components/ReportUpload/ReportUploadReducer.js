import Immutable from 'seamless-immutable';

import {
  REPORT_UPLOAD_PROCESS_START,
  REPORT_UPLOAD_PROCESS_STOP,
  REPORT_UPLOAD_PROCESS_FINISH,
} from './ReportUploadConstants';

const initialState = Immutable({
  /** insert ReportGenerate state here */
  processID: 0,
  status: 'stopped',
});

export default (state = initialState, action) => {
  const { payload: { processID, status } = {} } = action;
  switch (action.type) {
    case REPORT_UPLOAD_PROCESS_START:
      return state.merge({
        processID,
        status,
      });
    case REPORT_UPLOAD_PROCESS_STOP:
      return state.merge({
        status,
      });
    case REPORT_UPLOAD_PROCESS_FINISH:
      return state.merge({
        status,
      });
    default:
      return state;
  }
};
