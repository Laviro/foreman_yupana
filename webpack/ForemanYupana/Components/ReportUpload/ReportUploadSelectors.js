import { selectForemanYupana } from '../../ForemanYupanaSelectors';

export const selectReportUpload = state => selectForemanYupana(state).uploading;
export const selectProcessID = state => selectReportUpload(state).processID;
export const selectPollingProcessID = state =>
  selectReportUpload(state).pollingProcessID;
export const selectLogs = state => selectReportUpload(state).logs;
export const selectCompleted = state => selectReportUpload(state).completed;

export const selectLoading = state =>
  selectReportUpload(state).status === 'running';

export const selectExitCode = state => {
  const { status } = selectReportUpload(state);
  let exitCode;
  switch (status) {
    case 'success':
      exitCode = '0';
      break;
    case 'failure':
      exitCode = '1';
      break;
    default:
      exitCode = '--';
  }
  return exitCode;
};
