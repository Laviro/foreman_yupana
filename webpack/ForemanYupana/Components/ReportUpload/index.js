import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './ReportUploadActions';
import reducer from './ReportUploadReducer';

import ReportUpload from './ReportUpload';
import {
  selectLogs,
  selectProcessID,
  selectCompleted,
  selectLoading,
  selectExitCode,
  selectFiles,
  selectError,
} from './ReportUploadSelectors';

// map state to props
const mapStateToProps = state => ({
  /** add state keys here */
  processID: selectProcessID(state),
  logs: selectLogs(state),
  completed: selectCompleted(state),
  loading: selectLoading(state),
  exitCode: selectExitCode(state),
  files: selectFiles(state),
  error: selectError(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { uploading: reducer };

// export connected component
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ReportUpload);
