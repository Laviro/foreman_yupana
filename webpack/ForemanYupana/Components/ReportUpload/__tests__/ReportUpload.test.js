import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import ReportUpload from '../ReportUpload';

const fixtures = {
  'render without Props': {},
  /** fixtures, props for the component */
};

describe('ReportUpload', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(ReportUpload, fixtures));
});
