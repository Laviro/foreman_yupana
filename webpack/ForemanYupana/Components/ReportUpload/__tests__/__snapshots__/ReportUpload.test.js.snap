// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportUpload rendering render without Props 1`] = `
<TabContainer
  className="report-upload"
>
  <TabHeader
    exitCode={0}
    onRestart={[Function]}
  />
  <TabBody
    completed={0}
    loading={false}
    logs={
      Array [
        "No running process",
      ]
    }
  >
    No running process
  </TabBody>
  <TabFooter>
    <Tree
      files={Array []}
    />
    <FileDownload
      onClick={[Function]}
    />
  </TabFooter>
</TabContainer>
`;
